pkg/cloud/services/networking/securitygroups.go

```diff
@@ -147,51 +153,57 @@ func generateControlPlaneGroup(clusterName string) infrav1.SecurityGroup {
 					RemoteIPPrefix: "0.0.0.0/0",
 				},
 				{
-					Direction:      "ingress",
-					EtherType:      "IPv4",
-					PortRangeMin:   22,
-					PortRangeMax:   22,
-					Protocol:       "tcp",
-					RemoteIPPrefix: "0.0.0.0/0",
+					Direction:     "ingress",
+					EtherType:     "IPv4",
+					PortRangeMin:  2379,
+					PortRangeMax:  2380,
+					Protocol:      "tcp",
+					RemoteGroupID: remoteGroupIDSelf,
+				},
+				{
+					Direction:     "ingress",
+					EtherType:     "IPv4",
+					PortRangeMin:  10250,
+					PortRangeMax:  10250,
+					Protocol:      "tcp",
+					RemoteGroupID: remoteGroupIDSelf,
 				},
 			},
 			defaultRules...,
 		),
 	}
 }
 
-func generateGlobalGroup(clusterName string) infrav1.SecurityGroup {
-	secGroupName := fmt.Sprintf("%s-cluster-%s-secgroup-%s", secGroupPrefix, clusterName, globalSuffix)
+func (s *Service) generateWorkerGroup(clusterName string) infrav1.SecurityGroup {
+	secWorkerGroupName := fmt.Sprintf("%s-cluster-%s-secgroup-%s", secGroupPrefix, clusterName, workerSuffix)
+	secControlPlaneGroupName := fmt.Sprintf("%s-cluster-%s-secgroup-%s", secGroupPrefix, clusterName, controlPlaneSuffix)
```

Comment:
what is the dependency from the workersecgroup to the control plane sec group?

Reply:
The worker security group accept 10250/tcp from control plane security group. Please see line 206.

Reply:
Seems a bit implicit to just return an empty security group if the control plane group doesn't exist. I don't get the logic behind that. What do we want to achieve with that?

Reply:
I will fix to return error if the control plane security group does not exist.

Reply:
@sbueringer I fixed this as below:
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/555/files#diff-981e50f3b13da070546823dd24e18b32R182-R188

Reply:
Can we already return the securityGroup in line 61 (call to createSecurityGroupIfNotExists) and then just hand it over to generateWorkerGroup so we don't need the additional lookup?

Reply:
I made generateControlPlaneSecuriyGroup and generateWorkerGroup into generateDesiredSecGroups, and build security group names only once.